### DISCUSSION TITLE
Discord.py Load Balancer and concurrency?

### DISCUSSION THREAD

**Discussion by blastbeng** [Rapptz/discord.py#9434] [Q&A]
Hello,
is there any way to implement something like a Load Balancer or how to use concurrency in slash commands using discord.py?

My bot architecture is as follow

       CORE (rest APIs)  (4 Processes) =>  Nginx Load Balancer => Discord.py client

Actually when the client calls the load balancer, it decides which of the 4 Rest APIs processes to call. In this architecture, the discord.py does absolutely nothing, but almost everything is delegated to the rest APIs.

This is good but the problem is that when someone uses a slash command inside Discord.py client and another user uses another command, the bot responds with: "The application did not respond" until the first command has finished its execution.

I was thinking of running 4 discord.py processes, but how can i balance them with a load balancer? 
Actually if i do something like that, the bot is more responsive but it "bugs" itself, as i receive multiple "Interaction Not Found" or "This interaction has already been responded before", so this is not the actual solution I guess

I'm not a genius and my code is dirty, but you can see the bot code here:

[https://github.com/blastbeng/discord-tts-bot/blob/main/client/main.py](https://github.com/blastbeng/discord-tts-bot/blob/main/client/main.py)

It is not public, I'm just using it on my own server

The way i start the bot is that:
```
class MyClient(discord.Client):
    def __init__(self, *, intents: discord.Intents):
        super().__init__(intents=intents)
        self.tree = app_commands.CommandTree(self)


client.run(os.environ.get("BOT_TOKEN"))
```

and a simple slashcommand:

```
@client.tree.command()
@app_commands.rename(text='text')
@app_commands.describe(text="The sentence to repeat")
@app_commands.rename(voice='voice')
@app_commands.describe(voice="The voice to use")
@app_commands.autocomplete(voice=rps_autocomplete)
@app_commands.checks.cooldown(1, 5.0, key=lambda i: (i.user.id))
@app_commands.guilds(discord.Object(id=os.environ.get("GUILD_ID")))
async def speak(interaction: discord.Interaction, text: str, voice: str = "google"):
    """Repeat a sentence"""
    is_deferred=True
    load_message = None
    try:
      
        await interaction.response.defer(thinking=True, ephemeral=True)
        load_message = await display_loader(interaction, get_current_guild_id(interaction.guild.id))
        if interaction.user.voice and interaction.user.voice.channel:
            voice_client = get_voice_client_by_guildid(client.voice_clients, interaction.guild.id)
            await connect_bot_by_voice_client(voice_client, interaction.user.voice.channel, interaction.guild)

            if voice_client and not hasattr(voice_client, 'play') and voice_client.is_connected():
                await interaction.followup.send(utils.translate(get_current_guild_id(interaction.guild.id),"Retry in a moment, I'm initializing the voice connection..."), ephemeral = True)
            elif voice_client and not voice_client.is_playing():

                currentguildid = get_current_guild_id(interaction.guild.id)
                
                lang_to_use = utils.get_guild_language(currentguildid)

                if voice != "google":
                    voice = listvoices_api(language=lang_to_use, filter=voice)

                if voice is not None:
                    url = os.environ.get("API_URL")+os.environ.get("API_PATH_AUDIO")+"repeat/learn/"+urllib.parse.quote(str(text))+"/" + urllib.parse.quote(voice) + "/"+urllib.parse.quote(currentguildid)+ "/" + urllib.parse.quote(lang_to_use)
                    await do_play(voice_client, url, interaction, currentguildid)
                else:
                    await interaction.followup.send(utils.translate(get_current_guild_id(interaction.guild.id),"Voice not found. To list all available voices, use:") + " /listvoices", ephemeral = True)      
            else:
                await interaction.followup.send(utils.translate(get_current_guild_id(interaction.guild.id),"The bot is already talking or another user is already using another command.") +"\n" + utils.translate(get_current_guild_id(interaction.guild.id),"Please try again later or use stop command"), ephemeral = True)            
        else:
            await interaction.followup.send(utils.translate(get_current_guild_id(interaction.guild.id),"You must be connected to a voice channel to use this command"), ephemeral = True)
        await load_message.delete()
    except Exception as e:
        await send_error(e, interaction, from_generic=False, is_deferred=is_deferred, load_message=load_message)
```

Thanks

**Comment by AbstractUmbra** [comment]
My query directly is: why do your commands fail if one is already running? This isn't something discord.py does, as that isn't "async" in nature.
Commands can run asynchronously unless something in the usercode side of things blocks that - which would be why your commands are failing.

**Reply by blastbeng** [reply]
I have migrated from requests to requests_async and the problem seems gone.

[https://pypi.org/project/requests-async/](https://pypi.org/project/requests-async/)

**Reply by No767** [reply]
Still even moving towards requests_async, that package hasn't been updated since 4 years ago and would recommend migrating to AIOHTTP instead
